### PR TITLE
feat: OpenApi-generation from within eclipse (#509)

### PR DESCRIPTION
**Pull request by andreas-eternach**
* Added life-cycle-mapping for recognition by M2E
* Make BuildContext injectable by M2E in oder to detect if json-source
has been modified and a regeneration is required.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.4.x`, `4.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

insert support for Eclipse-M2E to be capable of running code-generation in eclipse

To close https://github.com/OpenAPITools/openapi-generator/issues/509

### CODE REVIEW

**Comment by andreas-eternach** [comment]
i have no clue why circleci is failing, any explanation is very welcome.

**Comment by wing328** [comment]
@andreas-eternach the CircleCI issue is not related to this PR. If you merge the latest master into this branch or rebase on the latest master, the issue should go away.

**Comment by wing328** [comment]
cc @jmini

File: modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java [hunk 1]
```diff
@@ -573,30 +599,39 @@ public void execute() throws MojoExecutionException {
             return;
         }
         adjustAdditionalProperties(config);
-        try {
+//        try {
             new DefaultGenerator().opts(input).generate();
+            
+            if (buildContext != null) {
+            	buildContext.refresh(new File(getCompileSourceRoot()));
```
[the remark this answers is not in the document]
Comment (andreas-eternach) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: should be fixed

[the remark this answers is not in the document]
Comment (andreas-eternach) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: should be fixed with second commit.

File: modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java [hunk 2]
```diff
@@ -573,30 +599,39 @@ public void execute() throws MojoExecutionException {
...
+            }
         } catch (Exception e) {
             // Maven logs exceptions thrown by plugins only if invoked with -e
             // I find it annoying to jump through hoops to get basic diagnostic information,
             // so let's log it in any case:
-            getLog().error(e);
+        	if (buildContext != null) {
+            	buildContext.addError(inputSpecFile, 0, 0, "unexpected error in Open-API generation", e);
```
[the remark this answers is not in the document]
Comment (andreas-eternach) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: should be fixed with second commit.

File: modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java [hunk 3]
```diff
@@ -573,30 +599,39 @@ public void execute() throws MojoExecutionException {
...
+        	}
+        	getLog().error(e);
```
[the remark this answers is not in the document]
Comment (andreas-eternach) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: should be fixed with second commit.

File: modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java [hunk 4]
```diff
@@ -573,30 +599,39 @@ public void execute() throws MojoExecutionException {
             return;
         }
         adjustAdditionalProperties(config);
-        try {
+//        try {
```
[the remark this answers is not in the document]
Comment (andreas-eternach) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: should be fixed with second commit.

File: modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (wing328) [reply modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java]: @andreas-eternach thanks for the quick fix. The change looks good to me.

**Comment by wing328** [comment]
@andreas-eternach thanks for the PR, which has been included in the v3.3.3 release: https://twitter.com/oas_generator/status/1062929948191510528

**Pull request merged**